### PR TITLE
Feature: Implementation of Select tool

### PR DESCRIPTION
**Pull request by hunkim98**
🚀 [Related Issue: #28 ]

### Preview

https://github.com/hunkim98/dotting/assets/57612141/0585b037-79f6-4d4d-b6df-c7959586fd03

<br/>

## Implementation of select tool (+undo functionality)

- **[🎨Component] Create functions to manage select area**

  - I added more attributes to `InteractionLayer` to manage select tool
    - `movingSelectingArea` is for tracking the selected Area when the user is dragging his/her mouse
    - `movingSelectedPixels` is for tracking the **colored** pixels in the moving selected area. This is later used for updating `selectedAreaPixels`
    - `selectedAreaPixels` is for tracking the **static** selected area. This is set when `onMouseUp` is called. 
  
  - I added functions to relay the selected area change information
    - `relaySelectingAreaToSelectedArea` is a function that relays the **selecting** area into a **selected** area. A **select-ING** area is set only when there was no selected area in the interaction layer. On the other hand a **select-ED** area is a previously selected area already in the interaction layer. The user is free to move the **select-ED** area if it is not `null`.
    - `relayMovingSelectedAreaToSelectedArea` is a function that relays **moving** selected area into a **selected** area. Remind that `movingSelectedArea` is a temporary attribute that exists only when the user is dragging the **selected** area. After user has finished interacting with the **selected** area, the information of the `movingsSelectedArea` is transferred into `selectedArea`.
    - In `relayMovingSelectedAreaToSelectedArea` a newly created `SelectAreaMoveAction` is recorded. This action records the erased and colored pixels, plus the selected area positions.

- **[🎨Component] Add `mouseDownWorldPos` and `mouseMoveWorldPos` to `Editor` class**

  - Record mouse down world position and mouse move world position
    - To transform the `selectedArea` correctly, I had to track the diff points of the mouse down and mouse move coordinates. The `mouseDownWorldPos` and `mouseMoveWorldPos` attributes will also come in handy when we improve our pixel drawing mechanism and our extend grid algorithm. (Somebody could try it out if they would like to?)

- **[🔗Other] Create a new action `SelectAreaMoveAction`**

  - A new action for recording selected area position change and pixel change
    - When undo is called upon a select area change interaction, a user will expect the pixels and also the previous selected area to be renderd. Thus, I created a new `SelectAreaMoveAction` that records not only the changed pixel colors but also the selected area world position changes. I also have made the brush tool set to `BrushTool.SELECT` to help users use the select tool for the `SelectAreaMoveAction`.

- **[🔗Other] Add additional helper functions in `utils/position.ts`**

  - `getDoesAreaOverlapPixelgrid` function
    - There was a need to detect if the user's selected area does not overlap with the current pixel grid. On such scenarios, the user's selected area will be null and the data stored in it will be gone.
    - This function detects if the selected area overlaps with the current pixel grid

  - `getIsPointInsideRegion` function
    - There was a need to detect if the user's mouse is in the selectedArea for moving the selectedArea.

  - Change naming of some functions
    - `convertSelectingAreaToPixelGridArea`

<br/>

## Notes

- Cursor styles for managing selected area seems necessary

<br/>

## Next Up?

- I will update some event handlers for selected area change, erase pixel change and so on.

### CODE REVIEW

**Comment by lerrybe** [comment]
How about `crosshair` style for the selection, and `grabbing` style for moving selectedArea?
[cursor - CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor)
스크린샷 2023-05-12 오후 11 00 40
스크린샷 2023-05-12 오후 11 02 44

**Comment by lerrybe** [comment]
https://github.com/hunkim98/dotting/assets/71599639/ee742e24-f444-46bc-9a9b-fe38c01ad04d

Report: When double-click pixels in the selected area, the pixel information inside seems to be initialized.

**Comment by hunkim98** [comment]
> report.mov 
> When double-click pixels in the selected area, the pixel information inside seems to be initialized. Is this part intended?

I did not notice that! I will take a look into the double click error. Thank you for your report🚀

Also, the cursor css you have suggested seems to be great. I believe we could set that later when our cursor styling finishes

**Comment by hunkim98** [comment]
@lerrybe I have fixed the error of double click erasing the selected area

**Review by lerrybe — APPROVED** [review]

**Review by Lee-Si-Yoon — APPROVED** [review]

File: src/components/Canvas/Editor.tsx [hunk 1]
```diff
@@ -1183,14 +1244,87 @@ export default class Editor extends EventDispatcher {
...
+        this.interactionLayer.setSelectedArea(null);
+        this.interactionLayer.setSelectingArea({
+          startWorldPos: this.mouseDownWorldPos,
+          endWorldPos: this.mouseDownWorldPos,
+        });
+      } else {
+        // we will move the selected area if the mouse is in the previous selected area
+        // remove the selecting area if it exists
+        this.interactionLayer.setSelectingArea(null);
+        const previousSelectedArea = this.interactionLayer.getSelectedArea()!;
```
[the remark this answers is not in the document]
Comment (hunkim98) [reply src/components/Canvas/Editor.tsx:1269]: Nice!

File: src/components/Canvas/Editor.tsx [hunk 2]
```diff
@@ -1183,14 +1244,87 @@ export default class Editor extends EventDispatcher {
...
+        const data = this.dataLayer.getData();
+        const rowCount = this.dataLayer.getRowCount();
+        const columnCount = this.dataLayer.getColumnCount();
+        const rowKeys = getRowKeysFromData(data);
+        const columnKeys = getColumnKeysFromData(data);
+        const sortedRowKeys = rowKeys.sort((a, b) => a - b);
+        const sortedColumnKeys = columnKeys.sort((a, b) => a - b);
+        const { includedPixelsIndices } = convertWorldPosAreaToPixelGridArea(
```
[the remark this answers is not in the document]
Comment (hunkim98) [reply src/components/Canvas/Editor.tsx:1277]: Nice one. I think we could reset make selectedArea null when includedPixelIndices is null👍

**Pull request merged**